### DISCUSSION TITLE
FTP_TLS connction error : invalid cert

### DISCUSSION THREAD

**Discussion by JMFUS** [micropython/micropython#13308] [RP2040 / Pico]
Hello and Happy New Year 2024

I'm writing this post because I need help (I'm a beginner on the subject)

I'm working on a Raspberry pi pico WH, and I wrote a program in micropython RP2 to control a system.
This Raspberry is installed on this system outside my home and communicates via Wifi
What I would like to do is be able to update my program on my Raspberry by OTA
For this I installed an FTP server on a Windows PC, using Filezilla Server 1.8.0
I would like to use FTP over TLS, and I have therefore created a certificate which, unless I am mistaken, is stored in users/toto/appdata/roaming/filezilla/trustedcerts.xml
I can connect to my FTP server via Filezilla Client from another PC, so I'm sure the port management on my router is OK
My micropython program is:
```
import network
import gc
import time
import usocket
import ussl as ssl
from ftplibtls import FTP_TLS

with open(cert_file_path, 'rb') as f:
cacert = f.read()
f.close()

ftp_conn = FTP_TLS(keyfile=None,certfile=cert_file_path,cert_reqs=ssl.CERT_REQUIRED,ca_certs=cacert,server_hostname=server_address)

print("Connecting to the FTP server...")
try:
ftp_conn.connect(server_address,port)
except Exception as e:
print("Error connecting to FTP server:", str(e))

try:
print("Connection established. Attempting authentication...")
ftp_conn.login(user=OST_user, passwd=OST_password)
except Exception as e:
raise e

try:
ftp_conn.prot_p()
except Exception as e:
raise e

ftp_conn.cwd("/")
```

The result is:

Traceback (most recent call last):
File "<stdin>", line 168, in <module>
File "<stdin>", line 166, in <module>
File "ftplibtls.py", line 78, in login
File "ftplibtls.py", line 101, in auth
File "ftplibtls.py", line 99, in auth
ValueError: invalid cert

On the FTP server, I have:

<12/30/2023 6:20:54 PM> FTP Session 32 192.168.0.254 [Response] 220-FileZilla Server 1.8.0
<12/30/2023 6:20:54 PM> FTP Session 32 192.168.0.254 [Response] 220-Please visit https://filezilla-project.org/
<30/12/2023 18:20:54> FTP Session 32 192.168.0.254 [Response] 220 Welcome to OST-FTP
<12/30/2023 6:20:54 PM> FTP Session 32 192.168.0.254 [Command] AUTH TLS
<12/30/2023 6:20:54 PM> FTP Session 32 192.168.0.254 [Response] 234 Using authentication type TLS.

then nothing

Can anyone give me some info?

Thanks in advance

**Comment by sophiedegran** [comment]
As I tell you in #9222 you need to have openssl installed to convert your certificate in DER format
Please read precedent post about TLS and go to the filezilla forum to know how to generate your certificate... (perharps filezilla can generate cert in DER format?)
dans tous les cas, bonne année !

**Comment by JMFUS** [comment]
Yes i tried to create a DER file with openSSL but the file created by FZ is an XML fileI have to search other solutions Bonne année 2024 😉Envoyé depuis mon appareil Galaxy
-------- Message d'origine --------De : sophiedegran ***@***.***> Date : 31/12/2023  16:56  (GMT+01:00) À : micropython/micropython ***@***.***> Cc : JMFUS ***@***.***>, Author ***@***.***> Objet : Re: [micropython/micropython] FTP_TLS connction error : invalid cert (Discussion #13308) 
As I tell you in #9222 you need to have openssl installed to convert your certificate in DER format
Please read precedent post about TLS and go to the filezilla forum to know how to generate your certificate... (perharps filezilla can generate cert in DER format?)
dans tous les cas, bonne année !
